File: src/main/scala/kim/dencka/scala/easy/LongestCommonPrefix.scala
package kim.dencka.scala.easy

object LongestCommonPrefix extends App {

  val array = Array("flower", "flow", "flight")

  def longestCommonPrefix(strs: Array[String]): String = {
    val starter = strs.headOption.getOrElse("")

    @annotation.tailrec
    def findCommon(str1: String, prefix: String): String =
      if (str1.startsWith(prefix)) prefix
      else if (str1.length > prefix.length) findCommon(str1.init, prefix)
      else if (str1.length == prefix.length) findCommon(str1.init, prefix.init)
      else if (str1.length < prefix.length) findCommon(str1, prefix.init)
      else ""

    strs.tail.foldLeft(starter) { (acc, elem) =>
      findCommon(elem, acc)
    }

  }

  println(longestCommonPrefix(array))

}

File: src/main/scala/kim/dencka/java/easy/LongestCommonPrefix.java
package kim.dencka.java.easy;

class LongestCommonPrefix {
    public static void main(String[] args) {
        System.out.println(longestCommonPrefix(new String[]{"flower", "flow", "flight"}));
        System.out.println(longestCommonPrefix(new String[]{"ab", "a"}));
    }

    public static String longestCommonPrefix(String[] strs) {
        StringBuilder res = new StringBuilder();
        String cur = strs[0];
        boolean check = true;
        for (int i = 0; i < cur.length(); i++) {
            for (String str : strs) {
                try {
                    check = check && (str.charAt(i) == cur.charAt(i));
                } catch (Exception e) {
                    check = false;
                    break;
                }
            }
            if (!check) break;
            res.append(cur.charAt(i));
        }
        return res.toString();
    }
}
